juju/juju
# Changes k8s watcher to use informers.

### Checklist

----

## Description of change
Moves k8swatcher to use shared informers over raw k8s watchers. Benefits here are listing, resuming and connection handling. Also will use a shared cache for fetching objects instead of putting extra load on the API server.

## QA steps
Bootstrap juju controller to microk8s. The following steps will then kill the kube-apiserver to mimic connection loss.
```
sudo kill $(pgrep kube-apiserver); microk8s.start
```
Check the logs to make sure we are not seeing any error messages with `k8s event watcher closed, restarting`

Next step is to manually destroy some workload pods and test the status is still being reported correctly in juju.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1859894
